Codegen: Parse deprecated signatures as a full FunctionSchema

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #74728
* #74727
* __->__ #82179

Deprecated signatures are currently "parsed" manually to find the
relative order of the argument names and all other information is
inferred from the aten schema for the non-deprecated overload.
However, this leads to problems if the argument names don't match or
if there are multiple candidates that match the ATen function call.

Instead, this makes the deprecated function a full FunctionSchema and
so the entire python signature comes solely from the deprecated
schema, with the `aten:` clause only used for the dispatch lambda call.

I have confirmed locally that there is no change to
`python_torch_functionsEverything.cpp`.

peterbell10 committed:
Codegen: Parse deprecated signatures as a full FunctionSchema

Deprecated signatures are currently "parsed" manually to find the
relative order of the argument names and all other information is
inferred from the aten schema for the non-deprecated overload.
However, this leads to problems if the argument names don't match or
if there are multiple candidates that match the ATen function call.

Instead, this makes the deprecated function a full FunctionSchema and
so the entire python signature comes solely from the deprecated
schema, with the `aten:` clause only used for the dispatch lambda call.

I have confirmed locally that there is no change to
`python_torch_functionsEverything.cpp`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Parse deprecated signatures as a full FunctionSchema"


Deprecated signatures are currently "parsed" manually to find the
relative order of the argument names and all other information is
inferred from the aten schema for the non-deprecated overload.
However, this leads to problems if the argument names don't match or
if there are multiple candidates that match the ATen function call.

Instead, this makes the deprecated function a full FunctionSchema and
so the entire python signature comes solely from the deprecated
schema, with the `aten:` clause only used for the dispatch lambda call.

I have confirmed locally that there is no change to
`python_torch_functionsEverything.cpp`.

[ghstack-poisoned]

peterbell10 committed:
Update on "Codegen: Parse deprecated signatures as a full FunctionSchema"


Deprecated signatures are currently "parsed" manually to find the
relative order of the argument names and all other information is
inferred from the aten schema for the non-deprecated overload.
However, this leads to problems if the argument names don't match or
if there are multiple candidates that match the ATen function call.

Instead, this makes the deprecated function a full FunctionSchema and
so the entire python signature comes solely from the deprecated
schema, with the `aten:` clause only used for the dispatch lambda call.

I have confirmed locally that there is no change to
`python_torch_functionsEverything.cpp`.

[ghstack-poisoned]

peterbell10 committed:
Add deprecations for inplace overloads on "Codegen: Parse deprecated signatures as a full FunctionSchema"


Deprecated signatures are currently "parsed" manually to find the
relative order of the argument names and all other information is
inferred from the aten schema for the non-deprecated overload.
However, this leads to problems if the argument names don't match or
if there are multiple candidates that match the ATen function call.

Instead, this makes the deprecated function a full FunctionSchema and
so the entire python signature comes solely from the deprecated
schema, with the `aten:` clause only used for the dispatch lambda call.

I have confirmed locally that there is no change to
`python_torch_functionsEverything.cpp`.

[ghstack-poisoned]

peterbell10 commented on tools/autograd/deprecated.yaml:
```diff
@@ -1,92 +1,134 @@
   aten: add(self, other, alpha)
 
-- name: add(Tensor self, Scalar alpha, Tensor other, *, Tensor out)
+- name: add_(Tensor(a!) self, Scalar alpha, Tensor other) -> Tensor(a!)
```
The old `load_deprecated_signatures` treated functional an inplace schemas as equivalent, but the matching is much more strict now so these have to be written out explicitly.

albanD commented on tools/autograd/gen_python_functions.py:
```diff
@@ -543,66 +512,89 @@ def signature_deprecated(
+                        return False
+
+            return len(schema.returns) == len(aten_schema.returns) and all(
+                a.type == b.type for a, b in zip(schema.returns, aten_schema.returns)
```
This is not checking that the annotations match?

peterbell10 replied:
Changed to check that returns match exactly.

peterbell10 committed:
Update on "Codegen: Parse deprecated signatures as a full FunctionSchema"


Deprecated signatures are currently "parsed" manually to find the
relative order of the argument names and all other information is
inferred from the aten schema for the non-deprecated overload.
However, this leads to problems if the argument names don't match or
if there are multiple candidates that match the ATen function call.

Instead, this makes the deprecated function a full FunctionSchema and
so the entire python signature comes solely from the deprecated
schema, with the `aten:` clause only used for the dispatch lambda call.

I have confirmed locally that there is no change to
`python_torch_functionsEverything.cpp`.

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

peterbell10 commented:
@pytorchbot merge -g
